crowbar_framework/app/controllers/api/restart_management_controller.rb

```diff
@@ -0,0 +1,147 @@
+  # beforehand
+  def restarts_get
+    # maybe we want to blacklist some barclamps from the restart list?
+    blacklist = []
+    barclamps = BarclampCatalog.members("openstack").keys.reject { |m| blacklist.include? m }
+    # Chef query search to find all nodes that have requires_restart attribute not empty
+    nodes_with_restarts = NodeObject.find("requires_restart:*")
+    @final_list = {}
+    nodes_with_restarts.each do |node|
+      @final_list[node[:fqdn]] = { alias: node.alias }
+      barclamps.each do |barclamp|
+        # skip if there is not attributes for the barclamp or the requires_restart is empty
+        # why node.node.fetch? Dont ask me, ask the chef creators why this works
+        # https://github.com/SUSE-Cloud/chef/blob/10-stable-suse/chef/lib/chef/node.rb#L187
+        requires_restart = node.node.fetch(
+          "crowbar_wall", {}
+        ).fetch(
+          "requires_restart", {}
+        ).fetch(
+           barclamp.to_s, {}
+        )
+        next if requires_restart.empty?
+        @final_list[node[:fqdn]].update(barclamp => requires_restart)
+      end
+    end
+
+    render json: @final_list
+  end
+
+  def restarts_post
+    params.require(:node)
+    params.require(:service)
+
+    node_name = params[:node]
+    service = params[:service]
+    barclamps = BarclampCatalog.members("openstack").keys
+
+    # Search for the exact node with the provided params. This should yield an unique node
+    node = NodeObject.find("name:#{node_name} AND requires_restart:#{service}").first
+
+    # validate that node exists
+    raise Crowbar::Error::NotFound if node.nil?
+
+    barclamps.each do |barclamp|
+      next unless node.key? :crowbar_wall
+      next unless node[:crowbar_wall].key? :requires_restart
+      next unless node[:crowbar_wall][:requires_restart].key? barclamp
+      next unless node[:crowbar_wall][:requires_restart][barclamp].key? service
+      node[:crowbar_wall][:requires_restart][barclamp].delete(service)
+    end
+
+    node.save
+
+    head :ok
+  end
+
+  def configuration_get
+    # If its a GET call, just show a list of the barclamps disallow_flag status
+    @final_list = Crowbar::DataBagConfig.get_or_create_databag_item(
+      "crowbar-config", "disallow_restart"
+    )
+    @final_list.raw_data.delete("id")
+    render json: @final_list.raw_data
+  end
+
+  def configuration_post
+    params.require(:disallow_restart)
+    params.require(:barclamp)
+
+    disallow_restart = params[:disallow_restart] == "true" ? true : false
+    barclamp = params[:barclamp]
+
+    # validate that the barclamp exists
+    raise Crowbar::Error::NotFound unless BarclampCatalog.members("openstack").include? barclamp
+
+    proposal = Proposal.where(barclamp: barclamp).first
+
+    # validate proposal is active as we dont allow to disable restarts for first barclamp runs
+    raise Crowbar::Error::BarclampNotActive if proposal.nil?
```

Comment:
I'm not convinced the check is fully fine. Let's assume we have a barclamp that supports multiple proposal. If one is active, then this will also block restart when applying a second proposal for the first time.

Reply:
+1